### ISSUE TITLE
rstan:::rstudio_stanc() update syntax or add message in the error for Stan version

### ISSUE THREAD

**Issue by ssp3nc3r** [stan-dev/rstan#821]
#### Summary:

`RStudio` calls ` rstan:::rstudio_stanc()` to check the syntax of Stan files, which relies on the version coinciding with `rstan`, which can lag Stan versions, and its new syntax, for necessary development reasons. This can lead to RStudio / `rstan:::rstudio_stanc()` throwing an error for syntactically correct Stan code if compiled under the latest version (using, for example, the package interface `cmdstanr`).

E.g., `reduce_sum` type syntax not currently recognized.

It is not intuitive that RStudio would not be agnostic to the package interface when we have multiple interfaces for different Stan versions. Would it make sense to either a) include in the error message from `rstan:::rstudio_stanc()` the version of Stan used for syntax checking or b) have this RStudio-specific function check syntax against the latest Stan, but flag what may not compile with rstan's current implementation?

#### Reproducible Steps:

From the [Stan User Guide, now in version 2.24](https://mc-stan.org/docs/2_24/stan-users-guide/reduce-sum.html):

```
functions {
  real partial_sum(int[] y_slice,
                   int start, int end,
                   vector x,
                   vector beta) {
    return bernoulli_logit_lpmf(y_slice | beta[1] + beta[2] * x[start:end]);
  }
}
data {
  int N;
  int y[N];
  vector[N] x;
}
parameters {
  vector[2] beta;
}
model {
  int grainsize = 1;
  beta ~ std_normal();
  target += reduce_sum(partial_sum, y,
                       grainsize,
                       x, beta);
}
```

#### Current Output:

RStudio (Version 1.3.1070)'s syntax checking produces this error:

```
> rstan:::rstudio_stanc("test.stan")
SYNTAX ERROR, MESSAGE(S) FROM PARSER:
Variable "partial_sum" does not exist.
 error in 'model1d1918bc1980_test' at line 20, column 34
  -------------------------------------------------
    18:   int grainsize = 1;
    19:   beta ~ std_normal();
    20:   target += reduce_sum(partial_sum, y,
                                         ^
    21:                        grainsize,
  -------------------------------------------------

Error in stanc(filename, allow_undefined = TRUE) : 
  failed to parse Stan model 'test' due to the above error.
```

#### Expected Output:

See the suggestion above.

#### RStan Version:

2.21.2

#### R Version:

4.02

#### Operating System:

OSX Catalina

**Comment by jgabry** [comment]
Yeah this is definitely a problem now that CmdStanR exists. We (and/or RStudio) will probably start getting bug reports about this. I'm not sure what the best solution is but @bgoodri would it be possible for `rstan:::rstudio_stanc()` to do the parsing using the latest stanc3 js build, since that's already available to RStan for checking whether models will parse with future Stan versions?

**Comment by bgoodri** [comment]
rstudio_stanc() just calls stanc(), which does the parsing under the
current (i.e. 2.21.x) C++ parser and then checks the future (i.e. 2.24.x)
parser. Since reduce_sum() is not in the 2.21 language, then you get this
error. We can't switch rstudio_stanc() to the javascript parser just yet
because it doesn't handle some things that are legal in 2.21. My guess is
that we just have to wait until rstan 2.24 is ready and even then, then new
parser may break RStudio's syntax checking.

**Comment by jgabry** [comment]
> My guess is that we just have to wait until rstan 2.24 is ready and even then, then new parser may break RStudio's syntax checking.

Why would it break RStudio's syntax checking if RStudio is just calling an RStan function that presumably isn't broken?

**Comment by bgoodri** [comment]
RStudio uses pattern matching on the error message to mark where in the
file it thinks a syntax error is occurring, and those patterns are probably
messed up. Also, the #include mechanism will throw the line numbers off.

**Comment by jgabry** [comment]
Ok this sounds like it's going to be a problem for basically every Stan user who uses R. It would seem prudent to start working on preventing these problems before they happen. I would really like to avoid the situation where suddenly with RStan 2.24 RStudio's Stan file functionality breaks for everyone. 

Do we need to talk to the RStudio people about changing their pattern matching? Or submit a PR to them? 

> Also, the #include mechanism will throw the line numbers off.

Why is that not already a problem? Did the #include mechanism change in the latest releases? If so, are you just resigned to the fact that it won't work properly or do you have an idea for what we can do?

**Comment by bgoodri** [comment]
More likely we can rewrite the error message so that its patterns continue
to match what RStudio was expecting based on previous versions of Stan. The
previous parser would give the line number of the parent Stan file and the
line number of the the #included Stan file plus a few lines of context
where the error occurred. Before we can call the new javascript Stan
parser, we have to make a temporary Stan character vector that copies the
#included Stan files into the parent Stan file using the C preprocessor
(which currently seems to be broken for a lot of people using the CRAN
version) because the javascript virtual machine cannot access the file
system. Thus, if there is any parser error in the combined Stan program,
the line numbers referred to in the error message will be off by the number
of lines in the previously #included Stan files and we would need to offset
the line numbers in the error message to at least refer to the correct line
number in the parent Stan file that the user is usually looking at.
